rfcs/0007-handling-corruption.md

```diff
@@ -0,0 +1,392 @@
+Updating corrupted keys with the same data has no effect.
+
+#### RocksDB
+RocksDB is an embeddable persistent key-value store written in C++ and used by
+multiple databases for their physical durability layer.
+
+##### Truncates Faulty WAL Frames
+Like SQLite and Pebble, RocksDB truncates all subsequent WAL entries when it
+finds corruption in the WAL. Upon discovering corruption, it logs a warning,
+and the database remains available.
+
+```
+[db/db_impl/db_impl_open.cc:1119] 000004.log: dropping 1482 bytes; Corruption: checksum mismatch
+```
+
+##### Prioritizes Availability
+Similar to Pebble, RocksDB returns an error for keys in the SST affected by
+corruption, such as `Corruption: block checksum mismatch: stored(context
+removed) = 1930575595, computed = 3064321270, type = 4 in rocksdb/000008.sst
+offset 1613 size 1621`. Despite this, the database remains available. Keys not
+affected by the corruption remain accessible, and new keys can be added to the
+database.
+
+For my research into RocksDB and Pebble, See  https://github.com/thrawn01/db-corruption
+
+## Goals
+The goal of this proposal is to develop a robust and flexible error reporting
+and corruption handling system. This system will be suitable for building both
+durability and availability systems on top of SlateDB.
+
+- Ensure the database remains available in the face of corruption.
+- Report corruption errors directly to the caller when corruption is detected.
+- Provide a corruption API that developers and operators can use to repair and
+  attempt recovery of corrupted data.
+- Enhance the robustness of the WAL implementation by attempting to repair and
+  recover as much of the WAL as possible, rather than truncating the remaining
+  entries.
+
+### Proposed Solution
+To provide support both Availability and Durability, I propose adding a callback
+function called `DBOptions.on_corruption` which is called when SlateDB encounters
+corruption during operation. The use of a call back allows developers to decide their
+desired behavior when corruption is detected. The return value of the callback
+then informs SlateDB of the users desire to abort the current operation with an
+error, or to continue operation attempting to skip the encountered corruption.
+
+To support developers who value availability, the `on_corruption` callback can
+be designed to log or notify the operator of the corruption, but otherwise
+remaining available.
+
+To support developers who value durability, the `on_corruption` callback can be
+designed to log or notify the operator of the corruption, then return an error,
+indicating the operators desire to make the database unavailable in order to 
+avoid losing data which may be recoverable.
+
+Additionally, the call back can be used to notify logical systems built on top
+of SlateDB that corruption exists and take appropriate action to repair and
+restore corrupted data from replicas or backups.
```

Comment:
this is an interesting idea, should we design the callback function with this in mind as a primary use case? it seems like the proposed callback method returns either an error or nothing, it may make sense to allow the callback to optionally return a "fixed" blob (or an empty blob to indicate "skip")

this would require the callback to support blocking operations, which may be an undesirable tradeoff but figured I'd bring it up

Reply:
I thought about this. Howver, It's not always possible to recover much about what was lost, IE: the key could be gone, or multiple key/blocks could be gone, etc...  the only thing we know for certain is that some unexpected data was encountered while scanning an SST file. The only real option a logical system can take is to rebuild the entire database. In order to support this, the logical system will likely partition the system where each partition is a separate SlateDB database which can be replicated and recovered in it's entirety. This both avoids write synchronization and makes recovery of a bad partition more manageable.